### PR TITLE
Replace nanomsg with raw unix domain sockets.

### PR DESCRIPTION
**Pull request by tyler-smith**
Instead of using the nanomsg pubsub protocol and Mangos library we are going to use a very simple library that simply streams data to the socket. This PR implements the socket library and uses it to replace nanomsg in our IPC system.

Previous we experienced dropped messages. With this new setup I have not seen any dropped messages and additionally have seen a significant increase in throughput; more than an order of magnitude.

Tested on MacOS, Ubuntu 20. Amazon Linux 2, and Windows.

Docs PR: https://github.com/ava-labs/avalanche-docs/pull/149

### CODE REVIEW

**Comment by tyler-smith** [comment]
@StephenButtolph and/or @danlaine This is ready for review when you can. Thanks.

Docs PR: https://github.com/ava-labs/gecko/pull/360

File: ipcs/socket/socket.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	s.connLock.RLock()
+	conns := make([]net.Conn, len(s.conns))
+	copy(conns, s.conns)
+	s.connLock.RUnlock()
+
+	// Write to each connection
+	var err error
+	for _, conn := range conns {
+		if _, err = conn.Write(msg); err != nil {
+			return err
```
Comment (danlaine) [inline ipcs/socket/socket.go:92]: might be worth returning a little more context about the error here.

e.g. `return fmt.Errorf("failed to write message to %s: %w", conn.RemoteAddr(), err)

File: ipcs/socket/socket.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+type Client struct {
+	net.Conn
+}
+
+// Recv waits for a message from the socket. It's guaranteed to either return a
+// complete message or an error
+func (c *Client) Recv() ([]byte, error) {
+	// Read length
+	var sz int64
+	if err := binary.Read(c.Conn, binary.BigEndian, &sz); err != nil {
```
Comment (danlaine) [inline ipcs/socket/socket.go:129]: Should we enforce a maximum read length here? Can an adversary make us allocate a huge byte array?

File: ipcs/socket/socket.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+			return nil, ErrTimeout
+		}
+		return nil, err
+	}
+
+	// Create buffer for entire message and read it all in
+	msg := make([]byte, sz)
+	if _, err := io.ReadFull(c.Conn, msg); err != nil {
+		if isTimeoutError(err) {
+			return nil, ErrTimeout
```
Comment (danlaine) [inline ipcs/socket/socket.go:140]: Would add more context here, too

e.g. `return nil, fmt.Errorf("read from %s timed out", c.Conn.RemoteAddr())`

File: ipcs/socket/socket_unix.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+// removeIfStaleUnixSocket takes in a path and removes it iff it is a socket
+// that is refusing connections
+func removeIfStaleUnixSocket(socketPath string) error {
+	// Ensure it's a socket; if not return without an error
+	if st, err := os.Stat(socketPath); err != nil || st.Mode()&os.ModeType != os.ModeSocket {
+		return nil
+	}
+
+	// Try to connect
+	conn, err := net.DialTimeout("unix", socketPath, 100*time.Millisecond)
```
Comment (danlaine) [inline ipcs/socket/socket_unix.go:62]: I'd move `100 * time.Millisecond` into a constant

**Comment by tyler-smith** [comment]
@danlaine Addressed all comments so far. Let me know you think is reasonable for the message size limit.

**Comment by danlaine** [comment]
@tyler-smith The max network message size is 1 << 21 so let's use that value here. Sound ok?

**Comment by tyler-smith** [comment]
Great, I missed that. That works for me.

**Comment by danlaine** [comment]
lgtm

**Comment by tyler-smith** [comment]
I just pushed changes to address 2 discovered issues:

1) When encountering an error on Send we were previously aborting the entire Send attempt and skipping any remaining connections even though they may not have issues.

2) When encountering a dead connection we were not removing it, causing dead connections to potentially build up and combined with issue #1 it would prevent any new connections from receiving messages.

File: ipcs/socket/socket.go (hunk 2 above)
Comment (StephenButtolph) [reply ipcs/socket/socket.go:149]: This will only read 8 bytes.

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**